torchmultimodal/models/mdetr.py

```diff
@@ -0,0 +1,699 @@
+                src_mask=mask,
+                src_key_padding_mask=src_key_padding_mask,
+                pos=pos,
+            )
+
+        if self.norm is not None:
+            output = self.norm(output)
+
+        return output
+
+
+class TransformerDecoder(nn.Module):
+    """
+    A transformer decoder.
+
+    Args:   decoder_layer (nn.Module): Module for an individual decoder layer.
+            num_layers (int): Number of decoder layers.
+            norm (Optional[nn.Module]): Normalization applied after last decoder layer.
+                Default: None
+            return_intermediate (bool): Whether to return intermediate decoder outputs.
+                Default: False
+
+    Inputs: tgt (Tensor): The sequence to the decoder layer.
+            memory (Tensor): The sequence from the last layer of the decoder.
+            tgt_mask (Optional[Tensor]) The mask for the tgt sequence. Default: None
+            memory_mask (Optional[Tensor]): The mask for the memory sequence.
+                Default: None
+            tgt_key_padding_mask (Optional[Tensor]): The mask for the tgt keys per batch.
+                Default: None
+            memory_key_padding_mask (Optional[Tensor]):
+            pos (Optional[Tensor]): Positional embeddings applied to Q and K
+                self-attention matrices. Default: None
+            query_pos (Optional[Tensor]): Positional embeddings applied to Q
+                cross-attention matrix. Default: None
+    """
+
+    def __init__(
+        self,
+        decoder_layer: nn.Module,
+        num_layers: int,
+        norm: Optional[nn.Module] = None,
+        return_intermediate: bool = False,
+    ):
+        super().__init__()
+        self.layers = _get_clones(decoder_layer, num_layers)
+        self.num_layers = num_layers
+        self.norm = norm
+        self.return_intermediate = return_intermediate
+
+    def forward(
+        self,
+        tgt: Tensor,
+        memory: Tensor,
+        tgt_mask: Optional[Tensor] = None,
+        memory_mask: Optional[Tensor] = None,
+        tgt_key_padding_mask: Optional[Tensor] = None,
+        memory_key_padding_mask: Optional[Tensor] = None,
+        pos: Optional[Tensor] = None,
+        query_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+        output = tgt
+
+        intermediate = []
+
+        for layer in self.layers:
+            output = layer(
+                output,
+                memory,
+                tgt_mask=tgt_mask,
+                memory_mask=memory_mask,
+                tgt_key_padding_mask=tgt_key_padding_mask,
+                memory_key_padding_mask=memory_key_padding_mask,
+                pos=pos,
+                query_pos=query_pos,
+            )
+            if self.return_intermediate:
+                intermediate.append(self.norm(output))
+
+        if self.norm is not None:
+            output = self.norm(output)
+            if self.return_intermediate:
+                intermediate.pop()
```

Comment:
why the pop?

Reply:
Yeah actually I don't think this is doing anything. Let me clean this part up